File: backend/app/api/v1/endpoints/projects.py
"""Project REST endpoints for listings, filters, trajectories, and history."""

from __future__ import annotations

from typing import Literal
from fastapi import APIRouter, Depends, HTTPException, Path, Query, status
from sqlalchemy.orm import Session

from backend.app.db.session import get_db
from backend.app.schemas.project_intelligence import ProjectRiskIntelligenceResponse
from backend.app.schemas.project_month import ProjectMonthObservationRead
from backend.app.schemas.projects import (
    FilterOptionsResponse,
    PaginatedProjectsResponse,
    ProjectCostRevisionsResponse,
    ProjectDetailResponse,
    ProjectScheduleExtensionsResponse,
    ProjectTrajectoryResponse,
    QuickSearchResult,
)
from backend.app.services.project_intelligence import ProjectIntelligenceService
from backend.app.services.project_service import ProjectService

router = APIRouter()

SortByFields = Literal[
    "report_month",
    "project_name",
    "project_code",
    "original_cost",
    "revised_cost",
    "cumulative_expenditure",
    "physical_progress",
    "approval_date",
    "original_completion_date",
    "revised_completion_date",
]
SortOrder = Literal["asc", "desc"]


@router.get(
    "/projects",
    response_model=PaginatedProjectsResponse,
    summary="List Projects with Filtering and Pagination",
    description="Retrieve paginated list of project observations supporting sector, state, agency, report_month, and search filters.",
)
def list_projects(
    page: int = Query(default=1, ge=1, description="Page number (1-indexed)"),
    page_size: int = Query(default=20, ge=1, le=100, description="Items per page (max 100)"),
    project_code: str | None = Query(default=None, description="Filter by project code (prefix or substring)"),
    report_month: str | None = Query(default=None, description="Exact report month, e.g. 2026-07"),
    sector: str | None = Query(default=None, description="Filter by sector"),
    state: str | None = Query(default=None, description="Filter by state"),
    agency: str | None = Query(default=None, description="Filter by implementing agency"),
    ministry: str | None = Query(default=None, description="Filter by ministry"),
    search: str | None = Query(default=None, description="Search by project name, project code, or agency"),
    sort_by: SortByFields = Query(default="report_month", description="Field to sort by"),
    sort_order: SortOrder = Query(default="desc", description="Sort direction (asc or desc)"),
    db: Session = Depends(get_db),
) -> PaginatedProjectsResponse:
    """Return paginated project observations."""
    service = ProjectService(db)
    return service.list_projects(
        page=page,
        page_size=page_size,
        project_code=project_code,
        report_month=report_month,
        sector=sector,
        state=state,
        agency=agency,
        ministry=ministry,
        search=search,
        sort_by=sort_by,
        sort_order=sort_order,
    )


@router.get(
    "/projects/filters/options",
    response_model=FilterOptionsResponse,
    summary="Get Dynamic Filter Options",
    description="Retrieve distinct non-null values for sectors, states, agencies, ministries, and report months currently available in the database.",
)
def get_filter_options(
    db: Session = Depends(get_db),
) -> FilterOptionsResponse:
    """Return distinct filter options."""
    service = ProjectService(db)
    return service.get_filter_options()


@router.get(
    "/projects/search/quick",
    response_model=list[QuickSearchResult],
    summary="Quick Search for Projects",
    description="Lightweight endpoint matching project name or project code for discovery and autocomplete UI components.",
)
def quick_search(
    q: str = Query(..., min_length=1, description="Search query string"),
    limit: int = Query(default=10, ge=1, le=50, description="Maximum results to return (max 50)"),
    db: Session = Depends(get_db),
) -> list[QuickSearchResult]:
    """Return quick search results."""
    service = ProjectService(db)
    return service.quick_search(query=q, limit=limit)


@router.get(
    "/projects/{project_code}",
    response_model=ProjectDetailResponse,
    summary="Get Project Detail by Project Code",
    description="Retrieve project-level summary metrics, observation span, and latest observation record.",
    responses={
        status.HTTP_404_NOT_FOUND: {"description": "Project code was not found in the database"},
    },
)
def get_project_detail(
    project_code: str = Path(..., description="Canonical source project identifier"),
    db: Session = Depends(get_db),
) -> ProjectDetailResponse:
    """Return project detail and latest observation."""
    service = ProjectService(db)
    return service.get_project_detail(project_code=project_code)


@router.get(
    "/projects/{project_code}/latest-snapshot",
    response_model=ProjectMonthObservationRead,
    summary="Get Latest Project Observation Snapshot",
    description="Retrieve the complete 31-field observation record from the latest available report month.",
    responses={
        status.HTTP_404_NOT_FOUND: {"description": "Project code was not found in the database"},
    },
)
def get_latest_snapshot(
    project_code: str = Path(..., description="Canonical source project identifier"),
    db: Session = Depends(get_db),
) -> ProjectMonthObservationRead:
    """Return complete latest observation record."""
    service = ProjectService(db)
    return service.get_latest_snapshot(project_code=project_code)


@router.get(
    "/projects/{project_code}/trajectory",
    response_model=ProjectTrajectoryResponse,
    summary="Get Chronological Project Trajectory",
    description="Retrieve full chronological timeline of monthly observations ordered by report_month ASC without artificial imputation.",
    responses={
        status.HTTP_404_NOT_FOUND: {"description": "Project code was not found in the database"},
    },
)
def get_trajectory(
    project_code: str = Path(..., description="Canonical source project identifier"),
    db: Session = Depends(get_db),
) -> ProjectTrajectoryResponse:
    """Return chronological project timeline."""
    service = ProjectService(db)
    return service.get_trajectory(project_code=project_code)


@router.get(
    "/projects/{project_code}/cost-revisions",
    response_model=ProjectCostRevisionsResponse,
    summary="Get Historical Cost Revisions",
    description="Retrieve chronological original, revised, and cumulative expenditure observations without subjective classification.",
    responses={
        status.HTTP_404_NOT_FOUND: {"description": "Project code was not found in the database"},
    },
)
def get_cost_revisions(
    project_code: str = Path(..., description="Canonical source project identifier"),
    db: Session = Depends(get_db),
) -> ProjectCostRevisionsResponse:
    """Return cost revision history."""
    service = ProjectService(db)
    return service.get_cost_revisions(project_code=project_code)


@router.get(
    "/projects/{project_code}/schedule-extensions",
    response_model=ProjectScheduleExtensionsResponse,
    summary="Get Historical Schedule & Milestone History",
    description="Retrieve chronological original and revised completion dates and start dates without manufactured delay classifications.",
    responses={
        status.HTTP_404_NOT_FOUND: {"description": "Project code was not found in the database"},
    },
)
def get_schedule_extensions(
    project_code: str = Path(..., description="Canonical source project identifier"),
    db: Session = Depends(get_db),
) -> ProjectScheduleExtensionsResponse:
    """Return schedule milestone history."""
    service = ProjectService(db)
    return service.get_schedule_extensions(project_code=project_code)


@router.get(
    "/projects/{project_code}/risk-intelligence",
    response_model=ProjectRiskIntelligenceResponse,
    summary="Get Comprehensive Project Risk Intelligence",
    description="Retrieve unified, typed project risk intelligence orchestrating canonical observations, production schedule-risk serving, model governance, signed drivers, and factual signals.",
    responses={
        status.HTTP_404_NOT_FOUND: {"description": "Project code was not found in the database"},
        status.HTTP_422_UNPROCESSABLE_ENTITY: {"description": "Project code is blank or invalid"},
    },
)
def get_project_risk_intelligence(
    project_code: str = Path(..., description="Canonical source project identifier"),
    db: Session = Depends(get_db),
) -> ProjectRiskIntelligenceResponse:
    """Return unified Project Risk Intelligence response."""
    clean_code = project_code.strip()
    if not clean_code:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="Project code cannot be blank or whitespace-only",
        )
    service = ProjectIntelligenceService(db)
    return service.get_project_risk_intelligence(project_code=clean_code)

File: frontend/src/api/projects.ts
/**
 * Project API Endpoints
 */

import { apiClient, API_BASE_URL } from "./client.ts";
import type {
  FilterOptionsResponse,
  PaginatedProjectsResponse,
  ProjectCostRevisionsResponse,
  ProjectDetailResponse,
  ProjectListQueryParams,
  ProjectMonthObservationRead,
  ProjectRiskIntelligenceResponse,
  ProjectScheduleExtensionsResponse,
  ProjectTrajectoryResponse,
  QuickSearchResult,
} from "@/types/project.ts";

export interface MonthlyObservationPoint {
  report_month: string;
  observations: number;
}

export interface MonthlyObservationFilters {
  sector?: string;
  agency?: string;
  state?: string;
  ministry?: string;
}

/**
 * List projects with pagination, sorting, and multi-field filtering.
 */
export async function fetchProjects(params: ProjectListQueryParams = {}): Promise<PaginatedProjectsResponse> {
  return apiClient<PaginatedProjectsResponse>(`${API_BASE_URL}/projects`, {
    params: {
      page: params.page ?? 1,
      page_size: params.page_size ?? 20,
      project_code: params.project_code,
      report_month: params.report_month,
      sector: params.sector,
      state: params.state,
      agency: params.agency,
      ministry: params.ministry,
      search: params.search,
      sort_by: params.sort_by,
      sort_order: params.sort_order,
    },
  });
}

/**
 * Retrieve dynamic filter options (sectors, states, agencies, ministries, months).
 */
export async function fetchFilterOptions(): Promise<FilterOptionsResponse> {
  return apiClient<FilterOptionsResponse>(`${API_BASE_URL}/projects/filters/options`);
}

/**
 * Fetch observation counts across multiple months, with optional taxonomy filtering.
 */
export async function fetchMonthlyObservations(
  months: string[],
  filters?: MonthlyObservationFilters
): Promise<MonthlyObservationPoint[]> {
  if (!months || months.length === 0) return [];
  const results = await Promise.all(
    months.map(async (month) => {
      try {
        const res = await fetchProjects({
          report_month: month,
          sector: filters?.sector || undefined,
          agency: filters?.agency || undefined,
          state: filters?.state || undefined,
          ministry: filters?.ministry || undefined,
          page: 1,
          page_size: 1,
        });
        return {
          report_month: month,
          observations: res.total ?? 0,
        };
      } catch {
        return {
          report_month: month,
          observations: 0,
        };
      }
    })
  );
  return results.sort((a, b) => a.report_month.localeCompare(b.report_month));
}

/**
 * Lightweight autocomplete/quick search for projects.
 */
export async function fetchQuickSearch(query: string, limit = 10): Promise<QuickSearchResult[]> {
  return apiClient<QuickSearchResult[]>(`${API_BASE_URL}/projects/search/quick`, {
    params: { q: query, limit },
  });
}

/**
 * Retrieve project high-level summary and latest observation by project code.
 */
export async function fetchProjectDetail(projectCode: string): Promise<ProjectDetailResponse> {
  return apiClient<ProjectDetailResponse>(`${API_BASE_URL}/projects/${encodeURIComponent(projectCode)}`);
}

/**
 * Retrieve complete 31-field latest observation snapshot.
 */
export async function fetchLatestSnapshot(projectCode: string): Promise<ProjectMonthObservationRead> {
  return apiClient<ProjectMonthObservationRead>(
    `${API_BASE_URL}/projects/${encodeURIComponent(projectCode)}/latest-snapshot`
  );
}

/**
 * Retrieve full chronological timeline of monthly observations.
 */
export async function fetchProjectTrajectory(projectCode: string): Promise<ProjectTrajectoryResponse> {
  return apiClient<ProjectTrajectoryResponse>(
    `${API_BASE_URL}/projects/${encodeURIComponent(projectCode)}/trajectory`
  );
}

/**
 * Retrieve historical original vs revised cost vs cumulative expenditure observations.
 */
export async function fetchCostRevisions(projectCode: string): Promise<ProjectCostRevisionsResponse> {
  return apiClient<ProjectCostRevisionsResponse>(
    `${API_BASE_URL}/projects/${encodeURIComponent(projectCode)}/cost-revisions`
  );
}

/**
 * Retrieve historical milestone schedule extensions and completion dates.
 */
export async function fetchScheduleExtensions(projectCode: string): Promise<ProjectScheduleExtensionsResponse> {
  return apiClient<ProjectScheduleExtensionsResponse>(
    `${API_BASE_URL}/projects/${encodeURIComponent(projectCode)}/schedule-extensions`
  );
}

/**
 * Retrieve unified Project Risk Intelligence response by canonical project code.
 */
export async function fetchProjectRiskIntelligence(
  projectCode: string
): Promise<ProjectRiskIntelligenceResponse> {
  return apiClient<ProjectRiskIntelligenceResponse>(
    `${API_BASE_URL}/projects/${encodeURIComponent(projectCode.trim())}/risk-intelligence`
  );
}

